src/util/psbt/mod.rs

```diff
@@ -88,6 +91,82 @@ impl PartiallySignedTransaction {
+        // 1) For non-p2sh non-segwit outputs, script code is the public key.
+        let sighash = if let Some(ref non_witness_utxo) = inp.non_witness_utxo {
+            let spent_outpoint = self.global.unsigned_tx.input[idx].previous_output;
+            if spent_outpoint.txid != non_witness_utxo.txid() {
+                return Err(Error::InvalidNonWitnessUtxo {
+                    prevout_txid: spent_outpoint.txid,
+                    non_witness_utxo_txid: non_witness_utxo.txid()
+                });
+            }
+            let script_pubkey =  &non_witness_utxo.output[spent_outpoint.vout as usize].script_pubkey;
+            if let Some(ref redeem_script) = inp.redeem_script {
+                if redeem_script.to_p2sh() != *script_pubkey {
+                    return Err(Error::InvalidWitnessScript{
+                        expected: script_pubkey.clone().into_bytes(),
+                        actual: redeem_script.to_p2sh().into_bytes(),
+                    });
+                }
+                self.global.unsigned_tx.signature_hash(
+                    idx, &redeem_script, sighash_type.as_u32()
+                )
+            } else {
+                self.global.unsigned_tx.signature_hash(
+                    idx, &script_pubkey, sighash_type.as_u32()
+                )
+            }
+        } else if let Some(ref witness_utxo) = inp.witness_utxo {
+            let script_pubkey= if let Some(ref redeem_script) = inp.redeem_script {
+                if redeem_script.to_p2sh() != witness_utxo.script_pubkey {
+                    return Err(Error::InvalidWitnessScript{
+                        expected: witness_utxo.script_pubkey.clone().into_bytes(),
+                        actual: redeem_script.to_p2sh().into_bytes(),
+                    });
+                }
+                redeem_script
+            } else {
+                &witness_utxo.script_pubkey
+            };
+
+            let amt = witness_utxo.value;
+            let mut sighash_cache = SigHashCache::new(&self.global.unsigned_tx);
+            if let Some(ref witness_script) = inp.witness_script {
+                if witness_script.to_v0_p2wsh() != *script_pubkey {
+                    return Err(Error::InvalidWitnessScript{
+                        expected: script_pubkey.clone().into_bytes(),
+                        actual: witness_script.to_p2sh().into_bytes(),
+                    });
+                }
+                sighash_cache.signature_hash(idx, &witness_script, amt, sighash_type)
+            } else if script_pubkey.is_v0_p2wpkh() {
+                // Indirect way to get script code
+                let builder = Builder::new();
+                let script_code = builder
+                .push_opcode(opcodes::all::OP_DUP)
+                .push_opcode(opcodes::all::OP_HASH160)
+                .push_slice(&script_pubkey[2..22])
+                .push_opcode(opcodes::all::OP_EQUALVERIFY)
+                .into_script();
```

Comment:
This can be simplified once https://github.com/rust-bitcoin/rust-bitcoin/pull/387 will be merged - just another reason to have that PR, @apoelstra

Reply:
Merged now :)